practical/ProcessMining/group2/alphaplusminer/alpha_plus_miner.py

```diff
@@ -0,0 +1,479 @@
+                    else:
+                        choice_pairs.append((activity1,activity2))
+
+    
+        pairs = causality_pairs
+
+        # We need to check the combinations too, while checking if the activities are independent in their respective sets
+     
+
+        i = 0
+        j = len(choice_pairs)
+
+        while i < j :
+            current_choice_pair = choice_pairs[i]
+            for pair in choice_pairs:
+                union = True
+           
+                if len(SortedSet(current_choice_pair).intersection(SortedSet(pair))) != 0:
+                    for e1 in pair:
+                        if union == False:
+                            break
+                        for e2 in current_choice_pair:
+                            if self.relations[e1][e2] != "#":  #check if the activities are independent
+                                union = False
+                                break
+                    if union :  # if not independent, create a new pair which is the union of the 2 activities eg  [a,d]  +  [a, f] -> [a, d, f]
+                        
+                        new_pair = SortedSet(current_choice_pair) | SortedSet(pair) 
+                   
+                      
+                        if tuple(new_pair) not in choice_pairs:
+                     
+                        
+                            choice_pairs.append(tuple(new_pair))
+                            j = j + 1
+                            
+                
+            i = i + 1
+
+      # Union
+        for pair_choices1 in choice_pairs:
+            for pair_choices2 in choice_pairs:
+                relation_between_pair = None
+                makePair = True
+
+                intersection = SortedSet(pair_choices1).intersection(pair_choices2)
+                pair_choices2 = SortedSet(pair_choices2)
+
+                if len(intersection) != 0:
+                    # remove intersection terms in the second pair
+                    for term in intersection:
+                        pair_choices2.discard(term)
+
+                if len(pair_choices2) == 0:
+                    continue
+                pair_choices2 = tuple(pair_choices2)
+
+                for activity1 in pair_choices1:
+                    if not makePair:
+                        break
+                    for activity2 in pair_choices2:
+                        relation = self.relations[activity1][activity2]
+                        if (relation_between_pair is not None
+                            and relation_between_pair != relation):
+                            makePair = False
+                            break
+                        else:
+                            relation_between_pair = relation
+                        if relation != "->":
+                            makePair = False
+                            break
+                if makePair is True:
+                    if relation_between_pair == "->":
```

Comment:
And all these hardcoded relations types should be really replaced for code quality and maintainability